tweet_recommendations/our_method.py

```diff
@@ -0,0 +1,286 @@
+            count_before = len(x)
+        x = x.dropna(subset=[self.embedding_name])
+        if self.verbose:
+            print(f"Dropped: {count_before - len(x)} rows.")
+        if self.verbose:
+            print('Building Graph')
+        g = self._build_base_graph(x)
+
+        if self.verbose:
+            print('Calculating hashtag embeddings.')
+        g = self._calculate_hashtag_embeddings(g)
+
+        if self.verbose:
+            print('Calculating popularity measures.')
+        g = self._calculate_hashtag_mean_retweets(g)
+        g = self._calculate_edge_weights(g)
+        self._G = self._calculate_pagerank(g)
+
+        hashtags = [{'hashtag': node, **self._G.nodes[node]}
+                    for node in self._G.nodes
+                    if self._G.nodes[node]['node_type'] == 'hashtag']
+        self._hashtags_df = pd.DataFrame(hashtags)
+
+    def embed_lemmas(self, texts_lemmas):
+        if not isinstance(texts_lemmas, pd.Series):
+            texts_lemmas = pd.Series(texts_lemmas)
+
+        def _embed_text(lemmas):
+            vectors = []
+            for word in lemmas:
+                try:
+                    vectors.append(self.embedding_model.word_vec(word.lower()))
+                except KeyError:
+                    continue
+            if len(vectors) > 0:
+                return np.asarray(vectors).mean(axis=0)
+            else:
+                return None
+        if self.verbose:
+            print("Embedding lemmas")
+            tqdm.tqdm.pandas()
+            return texts_lemmas.progress_apply(_embed_text)
+        else:
+            return texts_lemmas.apply(_embed_text)
+
+    def transform(self, x: Union[Tuple[Tuple[str, ...]], Tuple[str, ...],
+                                 Tuple[np.ndarray]],
+                  **transform_params):
+
+        x = list(x)
+        # tokenize and lemmatize text
+        if isinstance(x[0], str):
+            x = [get_wcrft2_results_for_text(txt) for txt in x]
+
+        # turn tokens to embeddings
+        if isinstance(x[0], list):
+            x = list(self.embed_lemmas(x))
+
+        # calculate mean tweet embeddings
+        if len(x[0].shape) > 1:
+            x = [emb.mean(axis=0) for emb in x]
+
+        hashtag_emb = np.asarray(self._hashtags_df[self.embedding_name].tolist())
+
+        if self.verbose:
+            print("Calculating similarities")
+        sim = self.embedding_similarity(np.asarray(x), hashtag_emb)
+        sim = ((sim - sim.min(axis=1).reshape(-1, 1))
+               / (sim.max(axis=1) - sim.min(axis=1)
+                  + 1e-8).reshape(-1,1))
+
+        pop = self._hashtags_df[self.popularity_measure].to_numpy()
+        pop = self.normalise(pop)
+        print(sim.shape, pop.shape)
+        sim_pop = ((1 - self.ratio) * sim + (self.ratio * pop))
+        result = []
+        if self.verbose:
+            print("Calculating ranking.")
+        for i in tqdm.tqdm(range(len(x)), disable=not self.verbose):

```

Comment:
`tqdm.tqdm` dla `range` nie będzie wiedziało, po ilu elementach iteruje (`range` jest generatorem), co sprawia, że sam komunikat jest mniej czytelny. Scastować `range` do listy, albo dać parametr do tego konstruktora `total=len(x)`.

Reply:
Hmm.. na pewno? Wydaje mi się że w pythonie 3 już to powinno bez problemu wiedzieć po ilu iteruje.
![image](https://user-images.githubusercontent.com/35629146/56475533-c2064a00-6489-11e9-8602-4e5513a5f963.png)

Reply:
To w takim razie może to zmienili, bo jak pamiętam, to był z tym zawsze problem.